SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/PartitionManager.cpp

```diff
@@ -0,0 +1,557 @@
+	}
+
+	static constexpr uint32 k_PartitionsReserveRange = 1024;
+
+	const FSubView& WorkerView;
+	const FSubView& SystemWorkerView;
+	const FSubView& PartitionView;
+	TUniquePtr<InterestFactory> InterestF;
+
+	Worker_EntityId StrategyWorkerEntityId = 0;
+	TOptional<Worker_RequestId> StrategyWorkerRequest;
+
+	TMap<Worker_RequestId_Key, FPartitionHandle> PartitionCreationRequests;
+	TOptional<Worker_RequestId> PartitionReserveRequest = 0;
+
+	TMap<Worker_EntityId_Key, FPartitionHandle> IdToPartitionsMap;
+	TSet<FPartitionHandle> Partitions;
+
+	TMap<Worker_EntityId_Key, ServerWorker> WorkersData;
+	TMap<Worker_EntityId_Key, Worker> SystemWorkersData;
+
+	Worker_EntityId FirstPartitionId = 0;
+	Worker_EntityId CurPartitionId = 0;
+
+	TSet<FLBWorkerHandle> ConnectedWorkers;
+	TArray<FLBWorkerHandle> ConnectedWorkersThisFrame;
+	TArray<FLBWorkerHandle> DisconnectedWorkersThisFrame;
+};
+
+FPartitionManager::~FPartitionManager() = default;
+
+FPartitionManager::FPartitionManager(Worker_EntityId InStrategyWorkerEntityId, ViewCoordinator& Coordinator,
+									 TUniquePtr<InterestFactory>&& InterestF)
+	: m_Impl(MakeUnique<Impl>(Coordinator, MoveTemp(InterestF)))
+{
+	m_Impl->StrategyWorkerEntityId = InStrategyWorkerEntityId;
+}
+
+void FPartitionManager::Init(ISpatialOSWorker& Connection)
+{
+	m_Impl->StrategyWorkerRequest = Connection.SendEntityCommandRequest(
+		m_Impl->StrategyWorkerEntityId, CreateClaimPartitionRequest(SpatialConstants::INITIAL_STRATEGY_PARTITION_ENTITY_ID),
+		SpatialGDK::RETRY_UNTIL_COMPLETE, {});
+
+	m_Impl->PartitionReserveRequest = Connection.SendReserveEntityIdsRequest(Impl::k_PartitionsReserveRange, RETRY_UNTIL_COMPLETE);
+}
+
+bool FPartitionManager::IsReady()
+{
+	return !m_Impl->StrategyWorkerRequest.IsSet();
+}
+
+TOptional<Worker_PartitionId> FPartitionManager::GetPartitionId(FPartitionHandle Partition)
+{
+	if (!m_Impl->Partitions.Contains(Partition))
+	{
+		return {};
+	}
+	FPartitionInternalState& PartitionState = *Partition->State;
+	if (!PartitionState.bAcked)
+	{
+		return {};
+	}
+	return PartitionState.Id;
+}
+
+FPartitionHandle FPartitionManager::CreatePartition(FString DisplayName, void* UserData, const QueryConstraint& Interest)
+{
+	FPartitionHandle NewPartition = MakeShared<FPartitionDesc>();
+	NewPartition->State = MakeUnique<FPartitionInternalState>();
+	NewPartition->State->UserData = UserData;
+	NewPartition->State->LBConstraint = Interest;
+	NewPartition->State->DisplayName = MoveTemp(DisplayName);
+
+	m_Impl->Partitions.Add(NewPartition);
+
+	return NewPartition;
+}
+
+void FPartitionManager::SetPartitionInterest(FPartitionHandle Partition, const QueryConstraint& NewInterest)
+{
+	if (!m_Impl->Partitions.Contains(Partition))
+	{
+		return;
+	}
+	FPartitionInternalState& PartitionState = *Partition->State;
+	PartitionState.LBConstraint = NewInterest;
+	PartitionState.bInterestDirty = true;
+}
+
+void FPartitionManager::AssignPartitionTo(FPartitionHandle Partition, FLBWorkerHandle Worker)
```

Comment:
This only gets called in legacy mode, is that intentional?

Reply:
Since the LegacyLB strategy is the only strategy implemented, yes.